lib/SILOptimizer/Mandatory/TFUtilities.cpp

```diff
@@ -1614,6 +1646,143 @@ GraphOperationDecoder::decodeAttributeName(Identifier name) {
   return { nameStr.substr(0, dollarLoc), opClass };
 }
 
+/// Analyze the array users of an _allocateUninitialized call, to see if they
+/// are all simple things we can remove.  If so, return add them all to
+/// arrayInsts and return false.  If so, return true.
+static bool analyzeArrayInitUses(SILValue v,
+                                 SmallPtrSet<SILInstruction*, 8> *arrayInsts) {
+  for (auto *use : v->getUses()) {
+    auto *user = use->getUser();
+
+    // We can always remove retain/release instructions.
+    if (isa<StrongRetainInst>(user) || isa<StrongReleaseInst>(user)) {
+      if (arrayInsts) arrayInsts->insert(user);
+      continue;
+    }
+
+    // We can look through unpacking and repacking of structs that eventually
+    // turn into a retain or release.
+    if (isa<StructExtractInst>(user) || isa<StructInst>(user)) {
+      if (arrayInsts) arrayInsts->insert(user);
+      if (analyzeArrayInitUses(user->getResults()[0], arrayInsts))
+        return true;
+      continue;
+    }
+  }
+  return false;
+}
+
+
+/// Given a SILValue that may be an array literal, attempt to decode it into the
+/// values that make up its elements.  If this fails or if the value is not an
+/// array, this returns a null Type.  Otherwise it decodes the array, returns
+/// the values of each element, and returns the element type of the array.
+///
+/// If arrayInsts is non-null and if decoding succeeds, this function adds
+/// all of the instructions relevant to the definition of this array into
+/// the set.  If decoding fails, then the contents of this set is undefined.
+Type GraphOperationInfo::
+decodeArrayElements(SILValue value, SmallVectorImpl<SILValue> &elements,
```

Comment:
should this function belong to TFConstExpr for more generality? e.g. #assert can then reason about the array elements.

Reply:
Yes, this is on my todo list.  I will move it as I figure out the right place for it.  This is a cut down (and more principled) version of the code used to analyze arrays after the performance optimizer.